### PR TITLE
Abolish Unittest

### PR DESCRIPTION
**Pull request by prescod**
Nothing for release notes

Abolish unittest.TestCase and replace it with pytest's simpler and more powerful syntax.

Beyond consistency, the other reason to do this is so that we can use fixtures in these classes.

### CODE REVIEW

**Comment by davisagli** [comment]
I'm excited about this, but I don't think all the tests are running:

on main:
2555 passed, 20 skipped, 494 warnings in 115.14s (0:01:55)
989 lines missing coverage

on this branch:
2390 passed, 20 skipped, 136 warnings in 108.25s (0:01:48)
1531 lines missing coverage

File: cumulusci/core/config/tests/test_config_expensive.py [hunk 1]
```diff
@@ -87,19 +86,19 @@ def test_load_universal_config_with_local(self, mock_class):
             expected_config = yaml.safe_load(f_expected_config)
         expected_config["tasks"]["newtesttask"] = {}
         expected_config["tasks"]["newtesttask"]["description"] = "test description"
-        self.assertEqual(config.config, expected_config)
+        assert config.config == expected_config
 
 
 @mock.patch("pathlib.Path.home")
```
Comment (davisagli) [inline cumulusci/core/config/tests/test_config_expensive.py:92]: When is a patch decorator on a class in effect? Should this be replaced with a fixture?

File: cumulusci/core/config/tests/test_config_expensive.py [hunk 2]
```diff
@@ -611,9 +607,9 @@ def test_create_org_command_error(self, Command):
             with open("tmp.json", "w") as f:
                 f.write("{}")
 
-            with self.assertRaises(ScratchOrgException) as ctx:
+            with pytest.raises(ScratchOrgException) as ctx:
                 config.create_org()
-                self.assertIn("scratcherror", str(ctx.error))
+                assert "scratcherror" in str(ctx.error)
```
[the remark this answers is not in the document]
Comment (davisagli) [reply cumulusci/core/config/tests/test_config_expensive.py:612]: (Maybe we should start measuring coverage of tests)

File: cumulusci/core/tests/test_flowrunner.py [hunk 3]
```diff
@@ -55,14 +54,14 @@ def _run_task(self):
 
 class AbstractFlowCoordinatorTest(object):
```
Comment (davisagli) [inline cumulusci/core/tests/test_flowrunner.py:55]:
```suggestion
class AbstractFlowCoordinatorTest:
```

File: cumulusci/core/tests/test_flowrunner.py [hunk 4]
```diff
@@ -55,14 +54,14 @@ def _run_task(self):
...
     @classmethod
-    def setUpClass(cls):
-        super(AbstractFlowCoordinatorTest, cls).setUpClass()
+    def setup_class(cls):
+        super(AbstractFlowCoordinatorTest, cls).setup_class()
```
Comment (davisagli) [inline cumulusci/core/tests/test_flowrunner.py:58]:
```suggestion
        super().setup_class()
```

File: cumulusci/core/tests/utils.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ def emit(self, record):
             self.release()
 
     def reset(self):
-        """Reset the handler in TestCase.setUp() to clear the msg list"""
+        """Reset the handler in TestCase.setup_method() to clear the msg list"""
```
Comment (davisagli) [inline cumulusci/core/tests/utils.py:38]: Doesn't need to be in this PR, but we can replace MockLoggingHandler entirely with pytest's caplog fixture

File: cumulusci/robotframework/tests/test_template_util.py [hunk 6]
```diff
@@ -1,9 +1,7 @@
-import unittest
-
 from cumulusci.core import template_utils
 
 
-class TemplateUtils(unittest.TestCase):
+class TemplateUtils:
```
Comment (davisagli) [inline cumulusci/robotframework/tests/test_template_util.py:4]: This doesn't have the right naming convention to be detected by pytest as a test class

File: cumulusci/utils/tests/test_fileutils.py [hunk 7]
```diff
@@ -158,7 +158,7 @@ def test_windows_path(self):
         abspath = "c:\\foo\\bar"
         with open_fs_resource(abspath) as f:
             if sys.platform == "win32":
-                assert "c:\\foo\\bar" == str(f.getsyspath()), str(f.getsyspath())
+                assert str(f.getsyspath()), str(f.getsyspath()) == "c:\\foo\\bar"
```
Comment (davisagli) [inline cumulusci/utils/tests/test_fileutils.py:161]:
```suggestion
                assert str(f.getsyspath()), str(f.getsyspath()) == r"c:\foo\bar"
```

**Comment by davisagli** [comment]
@prescod Here are the test classes that are not getting run:
```
cumulusci/core/tests/test_flowrunner.py::FullParseTestFlowCoordinator
cumulusci/core/tests/test_flowrunner.py::PreflightFlowCoordinatorTest
cumulusci/core/tests/test_flowrunner.py::SimpleTestFlowCoordinator
cumulusci/core/tests/test_flowrunner.py::StepSpecTest
cumulusci/core/tests/test_source.py::TestGitHubSource
cumulusci/robotframework/tests/test_template_util.py::TemplateUtils
cumulusci/salesforce_api/tests/test_metadata.py::BaseTestMetadataApi
cumulusci/tasks/github/tests/test_merge.py::TestMergeBranch
cumulusci/tasks/salesforce/tests/test_CreateCommunity.py
cumulusci/tasks/salesforce/tests/test_ListCommunities.py
cumulusci/tasks/salesforce/tests/test_ListCommunityTemplates.py
cumulusci/tasks/salesforce/tests/test_PublishCommunity.py
cumulusci/tasks/tests/test_datadictionary.py
```

In most cases because they don't follow the Test* pattern that pytest is looking for, I think. But I don't know yet why TestGitHubSource and TestMergeBranch are missing.

**Comment by prescod** [comment]
@davisagli Thanks! I was just doing that diff. :)

**Comment by prescod** [comment]
@davisagli 

"cannot collect test class 'TestGitHubSource' because it has a __init__ constructor "

I can fix that.

File: cumulusci/core/config/tests/test_config_expensive.py (hunk 1 above)
Comment (prescod) [reply cumulusci/core/config/tests/test_config_expensive.py:92]: It's the same as patching each method individually. Seems to work.

**Review by davisagli — APPROVED** [review]

**Pull request merged**